python-checks/src/main/java/org/sonar/python/checks/ChildAndParentExceptionCaughtCheck.java

```diff
@@ -57,22 +68,80 @@ private static void checkCaughtExceptions(SubscriptionContext ctx, Map<ClassSymb
       if (caughtExceptionsWithSameSymbol.size() > 1) {
-        PreciseIssue issue = ctx.addIssue(currentException, "Remove this duplicate Exception class.");
+        IssueWithQuickFix issue = (IssueWithQuickFix) ctx.addIssue(currentException, "Remove this duplicate Exception class.");
+        addQuickFix(issue, currentException);
         caughtExceptionsWithSameSymbol.stream().skip(1).forEach(e -> issue.secondary(e, "Duplicate."));
       }
-      PreciseIssue issue = null;
-      for (Map.Entry<ClassSymbol, List<Expression>> otherEntry : caughtExceptionsBySymbol.entrySet()) {
-        ClassSymbol comparedSymbol = otherEntry.getKey();
-        if (currentSymbol != comparedSymbol && currentSymbol.isOrExtends(comparedSymbol)) {
-          if (issue == null) {
-            issue = ctx.addIssue(currentException, "Remove this redundant Exception class; it derives from another which is already caught.");
-          }
-          addSecondaryLocations(issue, otherEntry.getValue());
-        }
+
+      var caughtParentExceptions = caughtExceptionsBySymbol.entrySet()
+        .stream()
+        .filter(entry -> entry.getKey() != currentSymbol && currentSymbol.isOrExtends(entry.getKey()))
+        .collect(Collectors.toList());
+
+      if (!caughtParentExceptions.isEmpty()) {
+        var issue = (IssueWithQuickFix) ctx.addIssue(currentException, "Remove this redundant Exception class; it derives from another which is already caught.");
+        addQuickFix(issue, currentException);
+
+        caughtParentExceptions.stream()
+          .map(Map.Entry::getValue)
+          .forEach(entries -> addSecondaryLocations(issue, entries));
       }
     });
   }
 
+  private static void addQuickFix(IssueWithQuickFix issue, Expression currentException) {
+    var quickFix = createQuickFix(currentException);
+    if (quickFix != null) {
+      issue.addQuickFix(quickFix);
+    }
+  }
+
+  private static List<String> collectNames(Expression expression) {
+    expression = Expressions.removeParentheses(expression);
+    if (expression.is(Tree.Kind.TUPLE)) {
+      var tuple = (Tuple) expression;
+      return tuple.elements()
+        .stream()
+        .map(ChildAndParentExceptionCaughtCheck::collectNames)
+        .flatMap(Collection::stream)
+        .collect(Collectors.toList());
+    } else if (expression.is(Tree.Kind.NAME)) {
+      var name = (Name) expression;
+      return List.of(name.name());
+    } else if (expression.is(Tree.Kind.QUALIFIED_EXPR)) {
+      var name = TreeUtils.tokens(expression)
+        .stream()
+        .map(Token::value)
+        .collect(Collectors.joining());
+      return List.of(name);
+    }
+    throw new IllegalArgumentException("Unsupported kind of tree element: " + expression.getKind().name());
+  }
+
+  private static PythonQuickFix createQuickFix(Expression currentException) {
+    try {
+      var currentExceptionName = collectNames(currentException).get(0);
```

Comment:
One thing I don't really like is that we have this method `collectNames` that can deal with tuples, names and qualified names, but here we really don't expect to have a tuple.
Similarly, on the second invocation on line 130 (when retrieving the list of other exceptions present), we only really expect to have a tuple, if my understanding is correct.
Maybe we could have a `collectNames` method that deals only with names and qualified names, and another which would be `collectNamesFromTuple` which would call the first one.

Reply:
done